src/wasm/wat-parser.cpp

```diff
@@ -378,6 +434,92 @@ Result<> parseDefs(Ctx& ctx,
 // Types
 // =====
 
+// heaptype ::= x:typeidx => types[x]
+//            | 'func'    => func
+//            | 'extern'  => extern
+template<typename Ctx>
+Result<typename Ctx::HeapTypeT> heaptype(Ctx& ctx, ParseInput& in) {
+  if (in.takeKeyword("func"sv)) {
+    RETURN_OR_OK(HeapType::func);
+  }
+  if (in.takeKeyword("any"sv)) {
+    RETURN_OR_OK(HeapType::any);
+  }
+  if (in.takeKeyword("extern"sv)) {
+    RETURN_OR_OK(HeapType::any);
+  }
+  if (in.takeKeyword("eq"sv)) {
+    RETURN_OR_OK(HeapType::eq);
+  }
+  if (in.takeKeyword("i31"sv)) {
+    RETURN_OR_OK(HeapType::i31);
+  }
+  if (in.takeKeyword("data"sv)) {
+    RETURN_OR_OK(HeapType::data);
+  }
+  if (in.takeKeyword("array"sv)) {
+    return in.err("array heap type not yet supported");
+  }
+  auto type = typeidx(ctx, in);
+  CHECK_ERR(type);
+  return *type;
+}
+
+// reftype ::= 'funcref'   => funcref
+//           | 'externref' => externref
+//           | 'anyref'    => anyref
+//           | 'eqref'     => eqref
+//           | 'i31ref'    => i31ref
+//           | 'dataref'   => dataref
+//           | 'arrayref'  => arrayref
+//           | '(' ref null? t:heaptype ')' => ref null? t
+template<typename Ctx>
+MaybeResult<typename Ctx::TypeT> reftype(Ctx& ctx, ParseInput& in) {
+  if (in.takeKeyword("funcref"sv)) {
+    RETURN_OR_OK(Type(HeapType::func, Nullable));
+  }
+  if (in.takeKeyword("externref"sv)) {
+    RETURN_OR_OK(Type(HeapType::any, Nullable));
+  }
+  if (in.takeKeyword("anyref"sv)) {
+    RETURN_OR_OK(Type(HeapType::any, Nullable));
+  }
+  if (in.takeKeyword("eqref"sv)) {
+    RETURN_OR_OK(Type(HeapType::eq, Nullable));
+  }
+  if (in.takeKeyword("i31ref"sv)) {
+    RETURN_OR_OK(Type(HeapType::i31, NonNullable));
+  }
+  if (in.takeKeyword("dataref"sv)) {
+    RETURN_OR_OK(Type(HeapType::data, NonNullable));
+  }
+  if (in.takeKeyword("arrayref"sv)) {
+    return in.err("arrayref not yet supported");
+  }
+
+  if (!in.takeSExprStart("ref"sv)) {
+    return {};
+  }
+
+  [[maybe_unused]] auto nullability =
+    in.takeKeyword("null"sv) ? Nullable : NonNullable;
+
+  auto type = heaptype(ctx, in);
+  CHECK_ERR(type);
+
+  if (!in.takeRParen()) {
+    return in.err("expected end of reftype");
+  }
+
+  if constexpr (parsingDecls<Ctx>) {
+    return Ok{};
+  } else if constexpr (parsingTypeDefs<Ctx>) {
+    return ctx.builder.getTempRefType(*type, nullability);
+  } else {
+    return Type(*type, nullability);
```

Comment:
Oh, actually the standard isorecursive types depend on the type already having been built, since the type use abbreviation (https://webassembly.github.io/spec/core/text/modules.html#abbreviations) that implicitly defines types first looks up whether the function type with the declared params and results has already been defined. That lookup requires the types to have already been built because the types need to have been deduplicated. So we should finish building the types at the end of phase 2 and use the type constructor directly for phase 3. For other nonstandard type systems we can add a constraint that explicit type definitions cannot refer to implicitly defined types to make this work.

Reply:
Then what if typeuses have types that are not defined in the type section, in which case we haven't seen the type?

Reply:
Isorecursive types are only allowed to refer to previously defined types or other types in the same recursion group. Implicitly defined types will always be in their own recursion group of size one, so they will only be allowed to refer to previously defined types. As long as we construct the implicitly defined types from first to last, all the of the types each one is allowed to refer to will already have been constructed when we parse it. 

(Actually, I guess an implicitly defined type could still refer to itself, so we might need a TypeBuilder to handle that, but it doesn't need to be the same TypeBuilder from phase 2)

Reply:
I guess I should study isorecursive types... But anyway, then is this not allowed in isorecursive type system?
```wast
(module
  (func $test (param i32) (result i32)
    ...
  )
)
```
This module doesn't have any `type` declarations but the function `test` has a function type use. Is this not parseable?

Reply:
> That module would parse ok. Phase two would parse zero explicit type definitions, then phase three would find the `(param i32) (result i32)` typeuse and add `[i32] -> [i32]` as the type at index 0. 

So in which case we first create types within `TypeBuilder` and which case we are allowed to create types using `Type` constructor on the fly?

Reply:
A TypeBuilder is only necessary to construct a mutually recursive group of types. Any type that only refers to other types that have already been built can be constructed with a constructor instead of a TypeBuilder.